### ISSUE TITLE
False non-primitive cast diagnostic

### ISSUE THREAD

**Issue by Veykril** [rust-lang/rust-analyzer#18626]
```rust

fn f() {
    trait T {
        fn f() {}
    }
    impl T for i32 {}
    impl T for u32 {}
    &[i32::f, u32::f] as &[fn()];
}

```
triggers two false diagnostics
```
non-primitive cast: `&[fn f<i32>(); 2]` as `&[fn()]`rust-analyzer[E0605](https://doc.rust-lang.org/stable/error_codes/E0605.html)
expected fn f<i32>(), found fn f<u32>()rust-analyzer[E0308](https://doc.rust-lang.org/stable/error_codes/E0308.html)
```

**Comment by ChayimFriedman2** [comment]
This appears to be a problem in Chalk, it assumes two different FnDefs (or Closures) cannot unify, while rustc coerces them to `fn()` if possible.

@Veykril Should we fix this now or wait for the new trait solver?

**Comment by Veykril** [comment]
Probably fine to wait, I noticed this because I introduced it in the r-a codebase now. But if no one else reports it its probably not high impact. Is it right that this is the on the trait solver here though and not us failing to do a correct coercion?

**Comment by ChayimFriedman2** [comment]
> Is it right that this is the on the trait solver here though and not us failing to do a correct coercion?

Yes. Chalk doesn't even check whether they're coercible, it just assumes they can unify only if they are exactly the same type. It's here:

https://github.com/rust-lang/chalk/blob/c83151f01e67af489d54423c71b9547cd80e8178/chalk-solve/src/infer/unify.rs#L266-L269

And a bit below for closures.

I also guess it will need some support in the chalk db to be able to determine if a closure is capture-free.

**Comment by ChayimFriedman2** [comment]
Oh wait! No, this isn't a problem in Chalk, my mistake. I will fix this soon.
